src/kernel/task.zig

```diff
@@ -99,6 +104,38 @@ pub const Task = struct {
         return task;
     }
 
+    pub fn createFromElf(program_elf: elf.Elf, kernel: bool, task_vmm: *vmm.VirtualMemoryManager(arch.VmmPayload), allocator: *Allocator) (bitmap.Bitmap(usize).BitmapError || vmm.VmmError || Allocator.Error)!*Task {
+        const task = try create(program_elf.header.entry_address, kernel, task_vmm, allocator);
+        errdefer task.destroy(allocator);
+
+        // Iterate over sections
+        var i: usize = 0;
+        errdefer {
+            // Free the previously allocated addresses
+            for (program_elf.section_headers) |header, j| {
+                if (j >= i)
+                    break;
+                if ((header.flags & elf.SECTION_ALLOCATABLE) != 0)
+                    task_vmm.free(header.virtual_address) catch |e| panic(null, "VMM failed to clean up a previously-allocated address after an error: {}\n", .{e});
+            }
+        }
+        while (i < program_elf.section_headers.len) : (i += 1) {
+            const header = program_elf.section_headers[i];
+            if ((header.flags & elf.SECTION_ALLOCATABLE) == 0) {
+                continue;
+            }
+            // If it is loadable then allocate it at its virtual address
+            const attrs = vmm.Attributes{ .kernel = kernel, .writable = (header.flags & elf.SECTION_WRITABLE) != 0, .cachable = true };
+            const vmm_blocks = std.mem.alignForward(header.size, vmm.BLOCK_SIZE) / vmm.BLOCK_SIZE;
+            const vaddr_opt = try task_vmm.alloc(vmm_blocks, header.virtual_address, attrs);
+            const vaddr = vaddr_opt orelse return if (try task_vmm.isSet(header.virtual_address)) error.AlreadyAllocated else error.OutOfBounds;
+            errdefer task_vmm.free(vaddr) catch |e| panic(@errorReturnTrace(), "Failed to free VMM memory in createFromElf: {}\n", .{e});
```

Comment:
This `errdefer` is what the larger `errdefer` does, so this can be removed. Else you will have a double free, but given the tests pass, then this double free is never hit, so either this is needed and I am dumb or there might need to be another test to check the cleanup.

Reply:
The double free is prevented with the `j >= i` check. The loop body can error *before* the virtual address is allocated (it can error on line 130) and so if this was merged with the errdefer block above it would try to free even though the address hadn't been allocated. I guess I could add an `if (task_vmm.isSet(...))` check to the errdefer block.

Reply:
Nah, I get you :)